### ISSUE TITLE
Hello, I have some questions about the loss function of the Gaussian distribution

### ISSUE THREAD

**Issue by SPOREIII** [arrigonialberto86/deepar#6]
The loss function of the Gaussian distribution given in the code is as follows:
```python
tf.reduce_mean(0.5*tf.math.log(sigma) + 0.5*tf.math.truediv(tf.math.square(y_true - y_pred), sigma)) + 1e-6 + 6  
```
I think it can be expressed as the following mathematical formula:
<img class="ee_img tr_noresize" alt="\frac{{\rm{1}}}{N} \times \sum {(\frac{1}{2} \times \log (\sigma ) + \frac{1}{2} \times \frac{{{{(z - \mu )}^2}}}{\sigma })}  + 1 \times {10^{ - 6}} + 6" src="http://chart.apis.google.com/chart?cht=tx&amp;chs=1x0&amp;chf=bg,s,FFFFFF00&amp;chco=000000&amp;chl=%5Cfrac%7B%7B%5Crm%7B1%7D%7D%7D%7BN%7D%20%5Ctimes%20%5Csum%20%7B%28%5Cfrac%7B1%7D%7B2%7D%20%5Ctimes%20%5Clog%20%28%5Csigma%20%29%20%2B%20%5Cfrac%7B1%7D%7B2%7D%20%5Ctimes%20%5Cfrac%7B%7B%7B%7B%28z%20-%20%5Cmu%20%29%7D%5E2%7D%7D%7D%7B%5Csigma%20%7D%29%7D%20%20%2B%201%20%5Ctimes%20%7B10%5E%7B%20-%206%7D%7D%20%2B%206" eeimg="1" style="vertical-align: middle; " longdesc="MathType!MTEF!2!1!+-
feaagKart1ev2aaatCvAUfeBSjuyZL2yd9gzLbvyNv2CaerbuLwBLn
hiov2DGi1BTfMBaeXatLxBI9gBaerbd9wDYLwzYbItLDharqqtubsr
4rNCHbWexLMBbXgBd9gzLbvyNv2CaeHbl7mZLdGeaGqiVu0Je9sqqr
pepC0xbbL8F4rqqrFfpeea0xe9Lq-Jc9vqaqpepm0xbba9pwe9Q8fs
0-yqaqpepae9pg0FirpepeKkFr0xfr-xfr-xb9adbaqaaeGaciGaai
aabeqaamaabaabauaakeaadaWcaaqaaiaabgdaaeaacaWGobaaaiab
gEna0oaaqaeabaGaaiikamaalaaabaGaaGymaaqaaiaaikdaaaGaey
41aqRaciiBaiaac+gacaGGNbGaaiikaiabeo8aZbbaaaaaaaaapeGa
aiyka8aacqGHRaWkdaWcaaqaaiaaigdaaeaacaaIYaaaaiabgEna0o
aalaaabaGaaiika8qacaWG6bGaeyOeI0IaeqiVd02daiaacMcadaah
aaWcbeqaaiaaikdaaaaakeaacqaHdpWCaaGaaiykaaWcbeqab0Gaey
yeIuoakiabgUcaRiaaigdacqGHxdaTcaaIXaGaaGimamaaCaaaleqa
baGaeyOeI0IaaGOnaaaakiabgUcaRiaaiAdaaaa!650E!
">
But according to the formula in the original text(DeepAR: Probabilistic Forecasting with Autoregressive Recurrent Networks), I derived the following result:
<img class="ee_img tr_noresize" alt=" - 1 \times \frac{{\rm{1}}}{N} \times \sum {(\log (\frac{1}{{\sqrt {2\pi } \sigma }}{e^{ - \frac{{{{(z - \mu )}^2}}}{{2{\sigma ^2}}}}}))} " src="http://chart.apis.google.com/chart?cht=tx&amp;chs=1x0&amp;chf=bg,s,FFFFFF00&amp;chco=000000&amp;chl=%20-%201%20%5Ctimes%20%5Cfrac%7B%7B%5Crm%7B1%7D%7D%7D%7BN%7D%20%5Ctimes%20%5Csum%20%7B%28%5Clog%20%28%5Cfrac%7B1%7D%7B%7B%5Csqrt%20%7B2%5Cpi%20%7D%20%5Csigma%20%7D%7D%7Be%5E%7B%20-%20%5Cfrac%7B%7B%7B%7B%28z%20-%20%5Cmu%20%29%7D%5E2%7D%7D%7D%7B%7B2%7B%5Csigma%20%5E2%7D%7D%7D%7D%7D%29%29%7D%20" eeimg="1" style="vertical-align: middle; " longdesc="MathType!MTEF!2!1!+-
feaagKart1ev2aaatCvAUfeBSjuyZL2yd9gzLbvyNv2CaerbuLwBLn
hiov2DGi1BTfMBaeXatLxBI9gBaerbd9wDYLwzYbItLDharqqtubsr
4rNCHbWexLMBbXgBd9gzLbvyNv2CaeHbl7mZLdGeaGqiVu0Je9sqqr
pepC0xbbL8F4rqqrFfpeea0xe9Lq-Jc9vqaqpepm0xbba9pwe9Q8fs
0-yqaqpepae9pg0FirpepeKkFr0xfr-xfr-xb9adbaqaaeGaciGaai
aabeqaamaabaabauaakeaacqGHsislcaaIXaGaey41aq7aaSaaaeaa
caqGXaaabaGaamOtaaaacqGHxdaTdaaeabqaaiaacIcaciGGSbGaai
4BaiaacEgacaGGOaWaaSaaaeaacaaIXaaabaWaaOaaaeaacaaIYaGa
eqiWdahaleqaaOGaeq4WdmhaaiaadwgadaahaaWcbeqaaiabgkHiTm
aalaaabaGaaiikaabaaaaaaaaapeGaamOEaiabgkHiTiabeY7aT9aa
caGGPaWaaWbaaWqabeaacaaIYaaaaaWcbaGaaGOmaiabeo8aZnaaCa
aameqabaGaaGOmaaaaaaaaaOWdbiaacMcapaGaaiykaaWcbeqab0Ga
eyyeIuoaaaa!5F15!
">
<img class="ee_img tr_noresize" alt=" = \frac{{\rm{1}}}{N} \times \sum {(\log (\sqrt {2\pi } \sigma ) + \frac{{{{(z - \mu )}^2}}}{{2{\sigma ^2}}})} " src="http://chart.apis.google.com/chart?cht=tx&amp;chs=1x0&amp;chf=bg,s,FFFFFF00&amp;chco=000000&amp;chl=%20%3D%20%5Cfrac%7B%7B%5Crm%7B1%7D%7D%7D%7BN%7D%20%5Ctimes%20%5Csum%20%7B%28%5Clog%20%28%5Csqrt%20%7B2%5Cpi%20%7D%20%5Csigma%20%29%20%2B%20%5Cfrac%7B%7B%7B%7B%28z%20-%20%5Cmu%20%29%7D%5E2%7D%7D%7D%7B%7B2%7B%5Csigma%20%5E2%7D%7D%7D%29%7D%20" eeimg="1" style="vertical-align: middle; " longdesc="MathType!MTEF!2!1!+-
feaagKart1ev2aaatCvAUfeBSjuyZL2yd9gzLbvyNv2CaerbuLwBLn
hiov2DGi1BTfMBaeXatLxBI9gBaerbd9wDYLwzYbItLDharqqtubsr
4rNCHbWexLMBbXgBd9gzLbvyNv2CaeHbl7mZLdGeaGqiVu0Je9sqqr
pepC0xbbL8F4rqqrFfpeea0xe9Lq-Jc9vqaqpepm0xbba9pwe9Q8fs
0-yqaqpepae9pg0FirpepeKkFr0xfr-xfr-xb9adbaqaaeGaciGaai
aabeqaamaabaabauaakeaacqGH9aqpdaWcaaqaaiaabgdaaeaacaWG
obaaaiabgEna0oaaqaeabaGaaiikaiGacYgacaGGVbGaai4zaiaacI
cadaGcaaqaaiaaikdacqaHapaCaSqabaGccqaHdpWCcaGGPaGaey4k
aSYaaSaaaeaacaGGOaaeaaaaaaaaa8qacaWG6bGaeyOeI0IaeqiVd0
2daiaacMcadaahaaWcbeqaaiaaikdaaaaakeaacaaIYaGaeq4Wdm3a
aWbaaSqabeaacaaIYaaaaaaakiaacMcaaSqabeqaniabggHiLdaaaa!5A4D!
">
<img class="ee_img tr_noresize" alt=" = \frac{{\rm{1}}}{N} \times \sum {(\log (\sqrt {2\pi } \sigma ) + \frac{{{{(z - \mu )}^2}}}{{2{\sigma ^2}}})} " src="http://chart.apis.google.com/chart?cht=tx&amp;chs=1x0&amp;chf=bg,s,FFFFFF00&amp;chco=000000&amp;chl=%20%3D%20%5Cfrac%7B%7B%5Crm%7B1%7D%7D%7D%7BN%7D%20%5Ctimes%20%5Csum%20%7B%28%5Clog%20%28%5Csqrt%20%7B2%5Cpi%20%7D%20%5Csigma%20%29%20%2B%20%5Cfrac%7B%7B%7B%7B%28z%20-%20%5Cmu%20%29%7D%5E2%7D%7D%7D%7B%7B2%7B%5Csigma%20%5E2%7D%7D%7D%29%7D%20" eeimg="1" style="vertical-align: middle; " longdesc="MathType!MTEF!2!1!+-
feaagKart1ev2aaatCvAUfeBSjuyZL2yd9gzLbvyNv2CaerbuLwBLn
hiov2DGi1BTfMBaeXatLxBI9gBaerbd9wDYLwzYbItLDharqqtubsr
4rNCHbWexLMBbXgBd9gzLbvyNv2CaeHbl7mZLdGeaGqiVu0Je9sqqr
pepC0xbbL8F4rqqrFfpeea0xe9Lq-Jc9vqaqpepm0xbba9pwe9Q8fs
0-yqaqpepae9pg0FirpepeKkFr0xfr-xfr-xb9adbaqaaeGaciGaai
aabeqaamaabaabauaakeaacqGH9aqpdaWcaaqaaiaabgdaaeaacaWG
obaaaiabgEna0oaaqaeabaGaaiikaiGacYgacaGGVbGaai4zaiaacI
cadaGcaaqaaiaaikdacqaHapaCaSqabaGccqaHdpWCcaGGPaGaey4k
aSYaaSaaaeaacaGGOaaeaaaaaaaaa8qacaWG6bGaeyOeI0IaeqiVd0
2daiaacMcadaahaaWcbeqaaiaaikdaaaaakeaacaaIYaGaeq4Wdm3a
aWbaaSqabeaacaaIYaaaaaaakiaacMcaaSqabeqaniabggHiLdaaaa!5A4D!
">
<img class="ee_img tr_noresize" alt=" = \frac{{\rm{1}}}{N} \times \sum {(\log (\sqrt {2\pi } ) + \log (\sigma ) + \frac{{{{(z - \mu )}^2}}}{{2{\sigma ^2}}})} " src="http://chart.apis.google.com/chart?cht=tx&amp;chs=1x0&amp;chf=bg,s,FFFFFF00&amp;chco=000000&amp;chl=%20%3D%20%5Cfrac%7B%7B%5Crm%7B1%7D%7D%7D%7BN%7D%20%5Ctimes%20%5Csum%20%7B%28%5Clog%20%28%5Csqrt%20%7B2%5Cpi%20%7D%20%29%20%2B%20%5Clog%20%28%5Csigma%20%29%20%2B%20%5Cfrac%7B%7B%7B%7B%28z%20-%20%5Cmu%20%29%7D%5E2%7D%7D%7D%7B%7B2%7B%5Csigma%20%5E2%7D%7D%7D%29%7D%20" eeimg="1" style="vertical-align: middle; " longdesc="MathType!MTEF!2!1!+-
feaagKart1ev2aaatCvAUfeBSjuyZL2yd9gzLbvyNv2CaerbuLwBLn
hiov2DGi1BTfMBaeXatLxBI9gBaerbd9wDYLwzYbItLDharqqtubsr
4rNCHbWexLMBbXgBd9gzLbvyNv2CaeHbl7mZLdGeaGqiVu0Je9sqqr
pepC0xbbL8F4rqqrFfpeea0xe9Lq-Jc9vqaqpepm0xbba9pwe9Q8fs
0-yqaqpepae9pg0FirpepeKkFr0xfr-xfr-xb9adbaqaaeGaciGaai
aabeqaamaabaabauaakeaacqGH9aqpdaWcaaqaaiaabgdaaeaacaWG
obaaaiabgEna0oaaqaeabaGaaiikaiGacYgacaGGVbGaai4zaiaacI
cadaGcaaqaaiaaikdacqaHapaCaSqabaGccaGGPaGaey4kaSIaciiB
aiaac+gacaGGNbGaaiikaiabeo8aZjaacMcacqGHRaWkdaWcaaqaai
aacIcaqaaaaaaaaaWdbiaadQhacqGHsislcqaH8oqBpaGaaiykamaa
CaaaleqabaGaaGOmaaaaaOqaaiaaikdacqaHdpWCdaahaaWcbeqaai
aaikdaaaaaaOGaaiykaaWcbeqab0GaeyyeIuoaaaa!5F58!
">
I don't know where I got it wrong and I cannot get the loss function form given in the code.

**Comment by benman1** [comment]
@SPOREIII I've changed this to your formula. Please check.

**Comment by arrigonialberto86** [comment]
Thanks for the question. This is a very old repo, but @benman1 did a great job cleaning up the mess recently. Regarding the loss function: the one previously used was wrong of course, and needs to be corrected according to what you are reporting above. As further discussion topic, we may consider using the new probabilistic layers introduced with Tensorflow Probability [here](https://www.tensorflow.org/probability/examples/Probabilistic_Layers_Regression) , where they use a simple `lambda` function to directly reference the negative log likelihood of a Gaussian distribution: `negloglik = lambda y, rv_y: -rv_y.log_prob(y)`. In this case we would not even need to express the likelihood explicitly
